### DISCUSSION TITLE
Non-CMake library distributed via vcpkg to CMake consumers - CMake config or Find module?

### DISCUSSION THREAD

**Discussion by thxkiwi** [microsoft/vcpkg#26286] [Q&A] [answered]
I write a library that is built with Visual Studio (msbuild).

I distribute binaries only of the library via vcpkg (I use a private vcpkg registry. The `portfile.cmake` for the port simply downloads a zip file of the binaries and headers and places them in the vcpkg install tree).

I have a client that is using CMake.

I can integrate vcpkg (manifest mode) into CMake and find_package() finds vcpkg ports except for mine.

I've arrived at the point that I think I need to distribute either a CMake config (`<my-package>-config.cmake`) or a CMake find module (`Find<my-package>.cmake`).

All of the books I've read for CMake seem to assume that you can easily export your CMake targets to create a CMake config. Since I don't build my library with CMake I'm not sure how to get a `<my-package>Targets.cmake`.

The documentation I'm reading on CMake find modules begins discussing resolving dependencies using `CMakeFindDependencyMacro` and `find_dependency()` but vcpkg also manages dependencies (this was the whole reason I distributed my library via vcpkg).

What's the solution here; CMake config or find module?

How does one write either if the project is not built with CMake?

**Comment by thxkiwi** [comment]
[Solution] find module based on https://gist.github.com/Osyotr

**Comment by ras0219-msft** [comment, marked as answer]
We recommend distributing a config modules, not a find module. `FindXYZ.cmake` is intended to be created by consumers, whereas `XYZ-config.cmake` is intended to be created by producers. You can only use `CMakeFindDependencyMacro` correctly from a config module; it cannot be used by a Find module.

While it's easy for CMake builds to generate config modules, they're still just CMake code that can be directly written. At a minimum, you can do exactly what you would do in a Find module. However, since you're the packager, you know more details about the build and can write even more correct code:
1. Instead of using `find_library()` (which could find libraries elsewhere on the system), you can directly point at the binaries: `${CMAKE_CURRENT_LIST_DIR}/.../../lib/mylib.lib`.
2. Instead of using `find_path()` (which could find headers elsewhere on the system), you can point directly at the headers: `${CMAKE_CURRENT_LIST_DIR}/../../include`
3. You can use `CMakeFindDependencyMacro` and `find_dependency()` to correctly find dependencies that you built with (if any)
